### PR TITLE
Add StreamingFileUpload.asInputStream

### PR DESCRIPTION
**Pull request by yawkat**
Repurpose some existing publisher->stream code to implement StreamingFileUpload.asInputStream. This will be useful for https://github.com/micronaut-projects/micronaut-object-storage/issues/113

### CODE REVIEW

**Comment by yawkat** [comment]
actually i'm going to make PublisherAsBlocking less abstract again.

File: http/src/main/java/io/micronaut/http/multipart/StreamingFileUpload.java [hunk 1]
```diff
@@ -79,4 +80,14 @@ default Publisher<Boolean> transferTo(OutputStream outputStream) {
...
     Publisher<Boolean> delete();
 
+    /**
+     * Create an {@link InputStream} that reads this file. The returned stream must be closed after
+     * use. The stream may block when data isn't yet available.
+     *
+     * @return An {@link InputStream} that reads this file's contents
+     * @since 4.2.0
+     */
+    default InputStream asInputStream() {
```
Comment (graemerocher) [inline http/src/main/java/io/micronaut/http/multipart/StreamingFileUpload.java:92]: add `@NonNull` to return type

Comment (graemerocher) [reply http/src/main/java/io/micronaut/http/multipart/StreamingFileUpload.java:92]: annotate with `@Blocking`

File: http-netty/src/main/java/io/micronaut/http/netty/PublisherAsBlocking.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+/**
+ * A subscriber that allows blocking reads from a publisher. Handles resource cleanup properly.
+ *
+ * @param <T> Stream type
+ * @since 4.2.0
+ * @author Jonas Konrad
+ */
+@Internal
+public final class PublisherAsBlocking<T> implements Subscriber<T>, Closeable {
```
Comment (graemerocher) [inline http-netty/src/main/java/io/micronaut/http/netty/PublisherAsBlocking.java:37]: does it need to be public?

File: http-netty/src/main/java/io/micronaut/http/netty/PublisherAsStream.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import java.io.InterruptedIOException;
+
+/**
+ * Transform a {@link PublisherAsBlocking} of buffers into a {@link InputStream}.
+ *
+ * @author Jonas Konrad
+ * @since 4.2.0
+ */
+@Internal
+public final class PublisherAsStream extends InputStream {
```
Comment (graemerocher) [inline http-netty/src/main/java/io/micronaut/http/netty/PublisherAsStream.java:33]: does it need to be public?

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: http-netty/src/main/java/io/micronaut/http/netty/PublisherAsBlocking.java (hunk 2 above)
Comment (yawkat) [reply http-netty/src/main/java/io/micronaut/http/netty/PublisherAsBlocking.java:37]: yes both used in two other packages

File: http/src/main/java/io/micronaut/http/multipart/StreamingFileUpload.java (hunk 1 above)
Comment (yawkat) [reply http/src/main/java/io/micronaut/http/multipart/StreamingFileUpload.java:92]: the method itself is not blocking even though the stream might be. should it still be annotated as such?

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/multipart/NettyPartData.java [hunk 4]
```diff
@@ -39,13 +38,13 @@
...
     private final Supplier<Optional<MediaType>> mediaTypeSupplier;
-    private final ThrowingSupplier<ByteBuf, IOException> byteBufSupplier;
+    private final Supplier<ByteBuf> byteBufSupplier;
 
     /**
      * @param mediaTypeSupplier The content type supplier
      * @param byteBufSupplier   The byte buffer supplier
      */
-    public NettyPartData(Supplier<Optional<MediaType>> mediaTypeSupplier, ThrowingSupplier<ByteBuf, IOException> byteBufSupplier) {
+    public NettyPartData(Supplier<Optional<MediaType>> mediaTypeSupplier, Supplier<ByteBuf> byteBufSupplier) {
```
Comment (alvarosanchez) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/multipart/NettyPartData.java:47]: Isn't this a breaking change?

Comment (yawkat) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/multipart/NettyPartData.java:47]: no the class is internal

Comment (alvarosanchez) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/multipart/NettyPartData.java:47]: ah, missed that, thanks

**Review by timyates — APPROVED** [review]

**Review by graemerocher — APPROVED** [review]

**Pull request merged**